synapse/storage/databases/main/end_to_end_keys.py

```diff
@@ -439,6 +443,100 @@ def _count_e2e_one_time_keys(txn: LoggingTransaction) -> Dict[str, int]:
             "count_e2e_one_time_keys", _count_e2e_one_time_keys
         )
 
+    async def count_bulk_e2e_one_time_keys_for_as(
+        self, user_ids: Collection[str]
+    ) -> TransactionOneTimeKeyCounts:
+        """
+        Counts, in bulk, the one-time keys for all the users specified.
+        Intended to be used by application services for populating OTK counts in
+        transactions.
+
+        Return structure is of the shape:
+          user_id -> device_id -> algorithm -> count
+        """
+
+        def _count_bulk_e2e_one_time_keys_txn(
+            txn: LoggingTransaction,
+        ) -> TransactionOneTimeKeyCounts:
+            user_in_where_clause, user_parameters = make_in_list_sql_clause(
+                self.database_engine, "user_id", user_ids
+            )
+            sql = f"""
+                SELECT user_id, device_id, algorithm, COUNT(key_id)
+                FROM devices
+                LEFT JOIN e2e_one_time_keys_json USING (user_id, device_id)
+                WHERE {user_in_where_clause}
+                GROUP BY user_id, device_id, algorithm
+            """
+            txn.execute(sql, user_parameters)
+
+            result: TransactionOneTimeKeyCounts = {}
+
+            for user_id, device_id, algorithm, count in txn:
+                device_count_by_algo = result.setdefault(user_id, {}).setdefault(
+                    device_id, {}
+                )
+                if algorithm is not None:
+                    # algorithm will be None if this device has no keys.
+                    device_count_by_algo[algorithm] = count
```

Comment:
```suggestion
                # algorithm will be None if this device has no keys.
                if algorithm is not None:
                    device_count_by_algo = result.setdefault(user_id, {}).setdefault(
                        device_id, {}
                    )
                    device_count_by_algo[algorithm] = count
```

Rather than creating empty dictionaries?

Reply:
I need to page in the MSC again (so good on you for mentioning, this could definitely use a comment), but I think we *want* the empty dicts to signal 'no keys'.

Reply:
Actually: the MSC says these can be omitted, but I think it's wrong. [I've opened a question to double check](https://github.com/matrix-org/matrix-doc/pull/3202/files#r807866130).